File: src/deeplecture/use_cases/dto/read_aloud.py
"""DTOs for Note Read-Aloud feature."""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class ReadAloudRequest:
    """Request to start read-aloud streaming."""

    content_id: str
    target_language: str  # ISO 639-1, e.g. "en", "zh"
    session_id: str
    source_language: str | None = None  # None = same as target (no translation)
    tts_model: str | None = None  # None = use task_models config
    start_paragraph: int = 0  # For paragraph-jump reconnect


@dataclass
class ParagraphMeta:
    """Metadata for a single paragraph."""

    index: int
    title: str | None
    sentence_count: int


@dataclass
class ReadAloudMeta:
    """First SSE event: total counts and paragraph metadata."""

    session_id: str
    total_paragraphs: int
    total_sentences: int
    paragraphs: list[ParagraphMeta] = field(default_factory=list)


@dataclass
class SentenceReady:
    """SSE event: a sentence's audio is ready for retrieval."""

    session_id: str
    variant_key: str
    paragraph_index: int
    sentence_index: int
    sentence_key: str  # e.g. "p0_s2" — used to fetch audio via REST
    original_text: str
    spoken_text: str  # May differ from original if translated
    cached: bool = False


@dataclass
class ReadAloudComplete:
    """SSE event: read-aloud generation finished."""

    session_id: str
    total_paragraphs: int
    total_sentences: int
    total_errors: int
    cancelled: bool = False

File: frontend/lib/api/readAloud.ts
/**
 * Read-Aloud APIs — SSE streaming + REST audio fetch for sentence-by-sentence TTS playback.
 *
 * Unlike task-based generation (podcast, cheatsheet), read-aloud uses a dedicated SSE channel
 * that pushes real-time signals as each sentence is synthesized. The frontend then fetches
 * individual sentence MP3 files via REST.
 */

import { api, API_BASE_URL } from "./client";

// ─── SSE Event Types ─────────────────────────────────────────

export interface ReadAloudMeta {
    sessionId: string;
    totalParagraphs: number;
    totalSentences: number;
    paragraphs: Array<{
        index: number;
        title: string | null;
        sentenceCount: number;
    }>;
}

export interface SentenceReady {
    sessionId: string;
    variantKey: string;
    paragraphIndex: number;
    sentenceIndex: number;
    sentenceKey: string;
    originalText: string;
    spokenText: string;
    cached?: boolean;
}

export interface ParagraphStart {
    paragraphIndex: number;
    title: string | null;
    sentenceCount: number;
}

export interface ReadAloudComplete {
    sessionId: string;
    totalParagraphs: number;
    totalSentences: number;
    totalErrors: number;
    cancelled?: boolean;
}

export interface ReadAloudError {
    error: string;
}

// ─── SSE Connection ──────────────────────────────────────────

export interface ReadAloudStreamParams {
    contentId: string;
    targetLanguage: string;
    sourceLanguage?: string;
    ttsModel?: string;
    startParagraph?: number;
}

/**
 * Create an EventSource for the read-aloud SSE stream.
 * The backend pushes events as each sentence is synthesized:
 *   read_aloud_meta → paragraph_start → sentence_ready* → paragraph_end → ... → read_aloud_complete
 */
export function createReadAloudEventSource(params: ReadAloudStreamParams): EventSource {
    const url = new URL(
        `${API_BASE_URL}/api/read-aloud/stream/${encodeURIComponent(params.contentId)}`
    );

    url.searchParams.set("target_language", params.targetLanguage);

    if (params.sourceLanguage) {
        url.searchParams.set("source_language", params.sourceLanguage);
    }
    if (params.ttsModel) {
        url.searchParams.set("tts_model", params.ttsModel);
    }
    if (params.startParagraph !== undefined && params.startParagraph > 0) {
        url.searchParams.set("start_paragraph", String(params.startParagraph));
    }

    return new EventSource(url.toString());
}

// ─── REST Audio Fetch ────────────────────────────────────────

/**
 * Build the URL for fetching a single sentence's MP3 audio.
 */
export function getSentenceAudioUrl(contentId: string, sentenceKey: string, variantKey: string): string {
    const url = new URL(
        `${API_BASE_URL}/api/read-aloud/audio/${encodeURIComponent(contentId)}/${encodeURIComponent(sentenceKey)}`
    );
    url.searchParams.set("variant_key", variantKey);
    return url.toString();
}

/**
 * Request cancellation for an active read-aloud session.
 */
export async function cancelReadAloud(contentId: string, sessionId: string): Promise<void> {
    await api.post(`/read-aloud/cancel/${encodeURIComponent(contentId)}`, {
        sessionId,
    });
}
